### PR TITLE
feat: allow customizable filename for file downloads

### LINKED ISSUE #869: Allow setting of filename for downloaded file via URL param
Right now it seems it's not possible to specify the file name that a downloaded file will have, i.e. I can't tell LAPIS to not use `details.json` as `filename` in content-disposition:

```
https://lapis.genspectrum.org/sample/details?limit=100&dataFormat=JSON&downloadAsFile=true
```

always responds with

Hyper 2024-07-23 13 38 47

`details.json` is very generic and not great when used by downstream websites that use LAPIS.

It should be easy to add a `filename=` URL param that simply changes the header response. That is, whatever is passed by `filename` URL param should be used in the header field `Content-Disposition-> filename=`

Right now, filenames seem to be hardcoded per endpoint, see https://github.com/GenSpectrum/LAPIS/blob/972a16c525c1ba8b9cdde06e2626dcda58c34f87/lapis/src/test/kotlin/org/genspectrum/lapis/controller/LapisControllerDownloadAsFileTest.kt#L186-L197

Here, one should allow a file to be specified via URL param (otherwise keep using this by default): https://github.com/GenSpectrum/LAPIS/blob/972a16c525c1ba8b9cdde06e2626dcda58c34f87/lapis/src/main/kotlin/org/genspectrum/lapis/controller/middleware/DownloadAsFileFilter.kt#L39-L60

Might be as simple as changing the value here, and using URL param if set: https://github.com/GenSpectrum/LAPIS/blob/972a16c525c1ba8b9cdde06e2626dcda58c34f87/lapis/src/main/kotlin/org/genspectrum/lapis/controller/middleware/DownloadAsFileFilter.kt#L42-L43

### PR DESCRIPTION
**Pull request by corneliusroemer**
Resolves #869

Proof of concept, would be great if someone with more experience with the code base could finish it by sorting out tests. Thanks!

How do you test these things other than using unit/e2e tests? Do you have a test deployment with some basic test data? That'd be cool :)

## PR Checklist
- [x] All necessary documentation has been adapted.
- [x] The implemented feature is covered by an appropriate test.

### CODE REVIEW

**Comment by corneliusroemer** [comment]
Thanks! That's more involved than I have time right now - if someone with a working dev setup could finish off tests that'd be great.

Reason I'm asking for test deployments is that they are easier to do exploratory testing on while one is figuring things out.

**Comment by fengelniederhammer** [comment]
I fixed a couple of things (wasn't working for POST requests) and added tests. Should be good now (other than squashing the commits).

**Review by JonasKellerer — APPROVED** [review]
LGTM in general.

I only found that when you select csv-without-headers, and compression/or without, and set the filename, then I get 'myFilename.json.zst'. But this seems to be another issue?

grafik

File: lapis/src/main/kotlin/org/genspectrum/lapis/openApi/Schemas.kt [hunk 1]
```diff
@@ -69,27 +69,38 @@ const val SEGMENT_SCHEMA = "Segment"
...
+const val DOWNLOAD_AS_FILE_DESCRIPTION =
+    """
 Set to true to make your browser trigger a download instead of showing the response content by setting the
 '$CONTENT_DISPOSITION' header to 'attachment'.
 """
 
-const val COMPRESSION_DESCRIPTION = """
+const val DOWNLOAD_FILE_BASENAME_DESCRIPTION =
+    """
+Specify download file basename, only useful with downloadAsFile set to true.
```
[the remark this answers is not in the document]
Comment (fengelniederhammer) [reply lapis/src/main/kotlin/org/genspectrum/lapis/openApi/Schemas.kt:93]: I changed it to 
```
Specify the download file basename, for example, specifying 'myFile' will result in a file named 'myFile.json' 
(when no compression is selected). 
This parameter only takes effect when 'downloadAsFile' is set to true.
```

**Comment by fengelniederhammer** [comment]
Thanks @corneliusroemer for your contribution! :) 

> I only found that when you select csv-without-headers, and compression/or without, and set the filename, then I get 'myFilename.json.zst'. But this seems to be another issue?

I created #871.

**Pull request merged**